source/opt/scalar_analysis.cpp

```diff
@@ -0,0 +1,386 @@
+    return instruction_map_[inst];
+
+  SENode* output = nullptr;
+  switch (inst->opcode()) {
+    case SpvOp::SpvOpPhi: {
+      output = AnalyzePhiInstruction(inst);
+      break;
+    }
+    case SpvOp::SpvOpConstant: {
+      output = AnalyzeConstant(inst);
+      break;
+    }
+    case SpvOp::SpvOpIAdd: {
+      output = AnalyzeAddOp(inst, false);
+      break;
+    }
+    case SpvOp::SpvOpISub: {
+      output = AnalyzeAddOp(inst, true);
+      break;
+    }
+    case SpvOp::SpvOpIMul: {
+      output = AnalyzeMultiplyOp(inst);
+      break;
+    }
+    default: {
+      output = CreateValueUnknownNode(inst);
+      instruction_map_[inst] = output;
+      break;
+    }
+  }
+  return output;
+}
+
+SENode* ScalarEvolutionAnalysis::AnalyzeConstant(const ir::Instruction* inst) {
+  if (inst->NumInOperands() != 1) {
+    assert(false);
+  }
+
+  int64_t value = 0;
+
+  // Look up the instruction in the constant manager.
+  const opt::analysis::Constant* constant =
+      context_->get_constant_mgr()->FindDeclaredConstant(inst->result_id());
+
+  if (constant->AsIntConstant()->type()->AsInteger()->IsSigned()) {
+    value = constant->AsIntConstant()->GetS32BitValue();
+  } else {
+    value = constant->AsIntConstant()->GetU32BitValue();
+  }
+
+  return CreateConstant(value);
+}
+
+// Handles both addition and subtraction. If the |sub| flag is set then the
+// addition will be op1+(-op2) otherwise op1+op2.
+SENode* ScalarEvolutionAnalysis::AnalyzeAddOp(const ir::Instruction* add,
+                                              bool sub) {
+  opt::analysis::DefUseManager* def_use = context_->get_def_use_mgr();
+  std::unique_ptr<SENode> add_node{new SEAddNode(this)};
+
+  add_node->AddChild(
+      AnalyzeInstruction(def_use->GetDef(add->GetSingleWordInOperand(0))));
+
+  SENode* op2 =
+      AnalyzeInstruction(def_use->GetDef(add->GetSingleWordInOperand(1)));
+
+  // To handle subtraction we wrap the second operand in a unary negation node.
+  if (sub) {
+    op2 = CreateNegation(op2);
+  }
+  add_node->AddChild(op2);
+
+  return GetCachedOrAdd(std::move(add_node));
+}
+
+SENode* ScalarEvolutionAnalysis::AnalyzePhiInstruction(
+    const ir::Instruction* phi) {
+  opt::analysis::DefUseManager* def_use = context_->get_def_use_mgr();
+
+  // Get the basic block this instruction belongs to.
+  ir::BasicBlock* basic_block =
+      context_->get_instr_block(const_cast<ir::Instruction*>(phi));
+
+  // And then the function that the basic blocks belongs to.
+  ir::Function* function = basic_block->GetParent();
+
+  // Use the function to get the loop descriptor.
+  ir::LoopDescriptor* loop_descriptor = context_->GetLoopDescriptor(function);
+
+  // We only handle phis in loops at the moment.
+  if (!loop_descriptor) return CreateCantComputeNode();
+
+  // Get the innermost loop which this block belongs to.
+  ir::Loop* loop = (*loop_descriptor)[basic_block->id()];
+  if (!loop) return CreateCantComputeNode();
```

Comment:
Seems like you should add a check that the loop has a latch or return can't compute here.